src/java.base/share/classes/java/nio/file/Path.java

```diff
@@ -248,6 +250,146 @@ public static Path of(URI uri) {
      */
     Path getFileName();
 
+    /**
+     * Returns the file name extension of this path as a {@code String},
+     * or a default value.  The extension is defined to be the portion of
+     * the {@code String} representation of the file name after the last
+     * dot ('.').  If the first character in the file name string is a dot
+     * it is ignored.  If the extension cannot be determined, then the
+     * parameter {@code defaultExtension} is returned.  This will occur if
+     * the path has zero elements ({@link #getFileName()} returns
+     * {@code null}), or the file name string does not contain a dot, only
+     * the first character is a dot, or the last character is a dot.
+     *
+     * @implSpec
+     * The default implementation is equivalent for this path to:
+     * <pre>{@code
+     *     String name = getFileName().toString();
+     *     int lastDot = name.lastIndexOf('.');
+     *     lastDot > 0 && lastDot < name.length() - 1 ?
+     *         name.substring(lastDot + 1) : defaultExtension;
+     * }</pre>
+     *
+     * @param   defaultExtension
+     *          the value to return if the extension is indeterminate;
+     *          may be {@code null}
+     *
+     * @return  the file name extension of this path, or
+     *          {@code defaultExtension} if the extension is indeterminate
```

Comment:
Seeing the use of `Optional<String> below`, returning an `Optional<String>` for this method might also be a good idea. (And dropping the `defaultExtension` parameter).
The testing for a missing extension is more obvious and the substitution for a missing extension can be handled by `Optional.orElse(`).

Reply:
I agree that returning an `Optional<String>` is a good idea. Getting rid of the parameter is a good thing. Commit 8ed8b1b25efa18a419d9176bff38f12616c9881d does this.